src/imbalances_script.py

```diff
@@ -0,0 +1,246 @@
+                return from_address, to_address, value
+            else:  # Withdrawal event
+                return from_address, None, value
+        except Exception as e:
+            print(f"Error decoding event: {str(e)}")
+            return None, None, None
+
+    def decode_sdai_event(self, event: Dict) -> Optional[Tuple[str, int]]:
+        """Decode sDAI event."""
+        try:
+            value_hex = event['data'][-30:]
+            
+            if isinstance(value_hex, bytes):
+                value = int.from_bytes(value_hex, byteorder='big')
+            else:
+                value = int(value_hex, 16)
+            
+            return SETTLEMENT_CONTRACT_ADDRESS, value
+        except Exception as e:
+            print(f"Error decoding sDAI event: {str(e)}")
+            return None
+
+    def find_chain_with_tx(self, tx_hash: str) -> Tuple[str, Web3]:
+        """
+        Find the chain where the transaction is present.
+        Returns the chain name and the web3 instance.
+        """
+        for chain_name, web3 in self.WEB3_INSTANCES.items():
+            if not web3.is_connected():
+                print(f"Could not connect to {chain_name}.")
+                continue
+            try:
+                web3.eth.get_transaction_receipt(tx_hash)
+                print(f"Transaction found on {chain_name}.")
+                return chain_name, web3
+            except Exception as e:
+                print(f"Transaction not found on {chain_name}: {e}")
+        raise ValueError(f"Transaction hash {tx_hash} not found on any chain.")
+
+    def get_transaction_receipt(self, tx_hash: str, web3: Web3) -> Optional[Dict]:
+        """
+        Get the transaction receipt from the provided web3 instance.
+        """
+        try:
+            return web3.eth.get_transaction_receipt(tx_hash)
+        except Exception as e:
+            print(f"Error getting transaction receipt: {e}")
+            return None
+
+    def process_event(self, event: Dict, inflows: Dict[str, int], outflows: Dict[str, int], address: str) -> None:
+        """Process a single event to update inflows and outflows."""
+        from_address, to_address, value = self.decode_event(event)
+        if from_address is None or to_address is None:
+            return
+        if to_address == address:
+            inflows[event['address']] = inflows.get(event['address'], 0) + value
+        if from_address == address:
+            outflows[event['address']] = outflows.get(event['address'], 0) + value
+
+    def process_sdai_event(self, event: Dict, imbalances: Dict[str, int], is_deposit: bool = True) -> None:
+        """Process an sDAI deposit or withdrawal event to update imbalances."""
+        decoded_event = self.decode_sdai_event(event)
+        if decoded_event is None:
+            return
+        to_address, value = decoded_event
+        if to_address == SETTLEMENT_CONTRACT_ADDRESS:
```

Comment:
One thing we observed with Felix is that this check is always satisfied, so it is not needed. However, I guess this means that you assume the settlement contract is always the one either receiving or sending away sDAI. See also line 118, inside the function `decode_sdai_event`.

I suspect this is not correct, in the sense that you shouldn't hardcode the settlement contract address there, although in most cases it will probably work fine. We should definitely understand if there is a bug here, and if so, fix it.

Reply:
Thanks for the catch. Since we're now reading txs from the db, there shouldn't be any issues of an external tx sdai values showing up. I want to test this with an sdai tx on uniswap or something to see what happens currently. 

Otherwise, I can make extract_events filter for settlement address in the first place instead of only having it in the calculate_imbalances call.